services/supervisor/src/lib.rs

```diff
@@ -123,6 +116,40 @@ fn update_configs(context: &mut CallContext<'_>, changes: Vec<ConfigChange>) ->
     Ok(())
 }
 
+/// Assigns the instance ID for a new service, initializing the schema `vacant_instance_id`
+/// entry if needed.
+fn assign_instance_id(context: &CallContext<'_>) -> InstanceId {
+    let mut schema = Schema::new(context.service_data());
+    match schema.assign_instance_id() {
+        Some(id) => id,
+        None => {
+            // Instance ID entry is not initialized, do it now.
+            // We have to do it lazy, since dispatcher doesn't know the amount
+            // of builtin instances until the genesis block is committed, and
+            // `before_commit` hook is not invoked for services at the genesis
+            // block.
+
+            // ID for the new instance is next to the highest builtin ID to avoid
+            // overlap if builtin identifiers space is sparse.
+            let dispatcher_schema = context.data().for_dispatcher();
+            let builtin_instances = dispatcher_schema.running_instances();
+
+            let new_instance_id = builtin_instances
+                .values()
+                .max_by_key(|spec| spec.id)
+                .map(|spec| spec.id)
+                .unwrap_or(SUPERVISOR_INSTANCE_ID)
+                + 1;
+
+            // We're going to use ID obtained above, so the vacant ID is next to it.
+            let vacant_instance_id = new_instance_id + 1;
+            schema.initialize_vacant_instance_id_entry(vacant_instance_id);
```

Comment:
Micro-nit / bikeshedding: I don't really see a purpose to extract a single assignment into a method...